### PR TITLE
Feature/edit track

### PR DESCRIPTION
**Pull request by tombu**
This branch introduces the long-awaited EDITOR !
It is an absolute basic implementation but I think it´s a good way to start with. The `BuildScreenStateManager` is now complete (and sadly a little to big for my taste) and can handle all transfers between the 4 (5) states: Drawing, Editing, Testing, Publishing (, Rasterizing)

I introduced an `Editor` class which creates the `EditView` (extends `TrackView`). The logic that happens in the `EditView` is not that complicated:
I extended the `drawTrack` method to fetch an array of the path points and iterate over it to draw a circle for each and bind a drag event listener. Inside the move event handler I update the point array and update the raphael path on the track. This way, `drawTrack` gets invoked because the raphael path changed.

My thoughts and issues:
- In the browser you can drag the handles outside the game window
- `Track`, `RaphaelPath` and `Path` are quite messy
- I couldn´t test it on the iPad
- Maybe the functionality of the `BuildScreen` should be re-thought to make it a little less fat

### CODE REVIEW

**Comment by stravid** [comment]
Awesometastic!

But I have to ruin the party.. on the iPad it doesn't work. I guess because the Raphael drag methods don't support touch events. Atleast it looks like that. But I could be totally wrong, the iPad reacts really weird..

I'm a little drunk so I can't verfiy this, but I think we have to implement our own drag methods with our custom events :(

File: app/assets/javascripts/slotcars/build/views/edit_view.js.coffee [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    @circles = @_paper.set()
+
+    for point, i in @pathPoints
+      circle = @_paper.circle point.x, point.y, 24
+      circle.attr stroke: '#000000', 'stroke-width': 3
+      circle.attr 'fill', '#00a2bf' # Raphael.getColor()
+      circle.data 'index', i
+      @circles.push circle
+
+    @circles.data 'viewContext', @
```
Comment (stravid) [inline app/assets/javascripts/slotcars/build/views/edit_view.js.coffee]: Please no single `@`, use `this` instead.

File: app/assets/javascripts/slotcars/build/views/edit_view.js.coffee [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    @circles.drag @_onEditHandleMove, @_onEditHandleDragStart
+
+  _onEditHandleDragStart: (x, y, event) ->
+    @dx = @dy = 0
+
+  _onEditHandleMove: (dx, dy, x, y, event) ->
+    X = (@attr 'cx') + (dx - (@dx || 0))
```
Comment (stravid) [inline app/assets/javascripts/slotcars/build/views/edit_view.js.coffee]: `deltaX` and `deltaY`

Is `cx` from Raphael?

**Comment by tombu** [comment]
Fortunately, we don´t have to implement an individual drag logic. I already figured out what the problem is: the paper gets completely cleared each time `drawTrack` is invoked. This means: all drag handles are removed and re-drawn.
Seems like the iPad does not like that. I´ll find a way to fix this but not tonight :wink:

Raphael has a drag example for touch devices: http://raphaeljs.com/touches.html

File: app/assets/javascripts/slotcars/build/views/edit_view.js.coffee (hunk 2 above)
Comment (tombu) [inline app/assets/javascripts/slotcars/build/views/edit_view.js.coffee]: Yeah, `cx` and `cy` is from Raphael. `c` stands for `center` ...

**Comment by DominikGuzei** [comment]
Hey Tom, this pull request is absolutely awesome!!! :-) Tried it today with Dave and it works like a charm on the laptop! The code looks pretty nice too and well-tested -> merge as soon as the iPAD issues are solved!

**Comment by tombu** [comment]
To fix the bug I had to re-structure `TrackView`. I introduced `updateTrack` which updates the `path` attributes of all raphael paths which are currently visible on the raphael paper. So as from now each raphael element should only be drawn once (they are stored in an array), the `track.raphaelPath` binding cares about updating the raphael paths on the paper.

**Comment by DominikGuzei** [comment]
someone tried it on the iPad?

**Comment by stravid** [comment]
It works on the iPad, though it feels a little bit slow and sluggish :(

Things Hannah noticed: You can drag an edit circle under the home button. Once there you can't change that :D

Should we still merge?

**Comment by tombu** [comment]
As slow and sluggish as drawing feels? Sadly, I have no iPad and therefore could just test it on the iPhone where it is also a bit sluggish. I hoped the performance would be better on the iPad.

I just introduced `Ember.run.snyc()` after updating the raphael path and at least on the iPhone it is less sluggish :smile:

**Comment by stravid** [comment]
I will test asap!

**Comment by tombu** [comment]
About Hannah´s notice: I recognized that, too. You can also drag the handle outside the game window on a desktop computer. How should we fix that? Should I limit the allowed coordinates for each circle?

**Comment by stravid** [comment]
That could be a solution for the desktop problem. For the home button thing, I'm not sure...

@MathiasPaumgarten @DominikGuzei @bmedicke what is your opinion?

**Comment by tombu** [comment]
Why not for the 'home button thing' ? We could allow a minimum `cy` of `120` ...

**Comment by stravid** [comment]
If you block the whole upper area I'm good with it. But just the button seemed a little odd :-)

**Comment by stravid** [comment]
It still feels sluggish. Maybe we should only show one layer when editing like in draw mode?

**Comment by DominikGuzei** [comment]
I would also use just one layer for the edit mode -> this is definitely faster than drawing 6 layers! I'm also ok with restricting the upper area of the screen, since the road should not cross the clock anyway (this could also be good for the drawing mode? but how? :-D)

**Comment by stravid** [comment]
In the drawing mode, and probably edit too, we could make the drawable area visible by putting a "sandbox" in there. Like we once said to make the build screen background look like sand and stuff we could use this plus the green to make a "box" where the users can draw the track.

**Comment by tombu** [comment]
I decided not to add a sandbox (and therefore no limitation of the dragging area) and prefer a discussion about a clean solution since we have that problem not just in the editing mode, but also in the drawing mode. Please @MathiasPaumgarten or @bmedicke test it on the iPad (@stravid has none at the moment) and merge if there are no objections.

**Comment by stravid** [comment]
We could clamp the drag `x` and `y` values. If you draw outside the sandbox the track is basically drawn as a straight line along the border.

I think this is better suited for a new pull request :-)

**Pull request merged**